crt0stack/src/lib.rs

```diff
@@ -925,4 +925,45 @@ mod tests {
         assert_eq!(reader.next(), None); // terminator
         assert_eq!(reader.next(), None); // don't overrun
     }
+
+    #[test]
+    #[should_panic]
+    fn crt0_stack_alignment() {
+        // Prepare the crt0 stack.
+        #[repr(C, align(32))]
+        struct Aligned(u8);
+        let mut crt0 = [0u8; 1024];
+        let crt0 = unsafe { crt0.align_to_mut::<Aligned>().1 };
+        let crt0 = unsafe { crt0.align_to_mut::<u8>().1 };
+        for i in 0..32 {
+            let crt0 = &mut crt0[i..];
+            let rand = unsafe { core::mem::transmute([0u8; 16]) };
+            //Set the arguments
+
+            let mut builder = Builder::new(crt0);
+            builder.push("/init").unwrap();
+
+            // Set the environment
+            let mut builder = builder.done().unwrap();
+            builder.push("LANG=C").unwrap();
+
+            // Set the aux vector
+            let mut builder = builder.done().unwrap();
+            builder.push(&Entry::ExecFilename("/init")).unwrap();
+            builder.push(&Entry::Platform("x86_64")).unwrap();
+            builder.push(&Entry::Uid(1000)).unwrap();
+            builder.push(&Entry::EUid(1000)).unwrap();
+            builder.push(&Entry::Gid(1000)).unwrap();
+            builder.push(&Entry::EGid(1000)).unwrap();
+            builder.push(&Entry::PageSize(4096)).unwrap();
+            builder.push(&Entry::Secure(false)).unwrap();
+            builder.push(&Entry::ClockTick(100)).unwrap();
+            builder.push(&Entry::Flags(0)).unwrap(); // TODO: https://github.com/enarx/enarx/issues/386
+            builder.push(&Entry::HwCap(0)).unwrap(); // TODO: https://github.com/enarx/enarx/issues/386
+            builder.push(&Entry::HwCap2(0)).unwrap(); // TODO: https://github.com/enarx/enarx/issues/386
+            builder.push(&Entry::Random(rand)).unwrap();
+
+            let _handle = builder.done().unwrap();
```

Comment:
```suggestion
            let handle = builder.done().unwrap();
            assert_eq!(handle.start_ptr() as usize % 32, 0);
```

Reply:
Following this suggestion: I think this would require the removal of the `#[should_panic]` attribute at the top of this unit test.

Reply:
@connorkuehl Correct. I thought I had added a suggestion to that effect, but it didn't make it in the review for some reason.

Reply:
Thank you @npmccallum and @connorkuehl for the detailed suggestions.

@npmccallum, Lily and I  were discussing the other day about the pointer returned by start_ptr() function. 
`handle.start_ptr()` is returning a &(), resulting in an error when I cast it to usize.
I tried using `*(&handle.0[handle.1])` instead, which is always returning 1. 

@npmccallum and @connorkuehl, since we are misaligning the stack, `handle.start_ptr() as usize % 32` would never be 0. Our test would be considered a success only if it panics right?
Shouldn't we have  `#[should_panic]` at the top?

Reply:
> @npmccallum, Lily and I were discussing the other day about the pointer returned by start_ptr() function.
> `handle.start_ptr()` is returning a &(), resulting in an error when I cast it to usize.
> I tried using `*(&handle.0[handle.1])` instead, which is always returning 1.

You probably just need to cast it to a pointer first. So: `handle.start_ptr() as *() as usize`

> @npmccallum and @connorkuehl, since we are misaligning the stack, `handle.start_ptr() as usize % 32` would never be 0.

This about this process as inputs and output. The inputs are:
1. a buffer of memory that will be used as the crt0 stack
2. arguments
3. environment variables
4. auxiliary vectors

The process writes inputs 2-4 into the buffer of memory (1) and returns a pointer to the "bottom" of the data written.

What we want to ensure is that no matter how the buffer of memory (1) is aligned, the output pointer is always correctly aligned.

> Our test would be considered a success only if it panics right?
> Shouldn't we have `#[should_panic]` at the top?

`assert_eq!(handle.start_ptr() as *() as usize % 32, 0);` verifies that the stack is properly aligned and panics otherwise. Therefore, the test will not panic when it is successful.

As a note: I think the actual stack alignment that we need is 16 bytes. But We're starting at 32 bytes to make sure the test fails and inspect the results.

Reply:
Thank you @npmccallum
`(handle.start_ptr() as *const ()` worked.
Do you think the correct stack alignment is 8 instead?
`(handle.start_ptr() as *const () as usize) % 32` always returns 8.
When I changed ` #[repr(C, align(32))]` --> ` #[repr(C, align(8))]` and 
`handle.start_ptr() as *const () as usize) % 32,` --> `handle.start_ptr() as *const () as usize) % 8`, the test passes.

Reply:
8 does not seem correct to me. @haraldh What are your thoughts here?

I'm rather surprised that it always returns 8.